### PR TITLE
Add new filters

### PR DESCRIPTION
**Pull request by apmypb**
Added filters that could replace Savitzky-Golay and a moving average window filter.
I wrote some working examples in the doc strings of the filters.
I can add tests if needed

### CODE REVIEW

**Comment by fhagemann** [comment]
Can you please also include code examples in this PR?

File: src/LegendDSP.jl [hunk 1]
```diff
@@ -40,12 +40,15 @@ include("saturation.jl")
 include("interpolation.jl")
 include("intersect_maximum.jl")
 include("dsp_sipm.jl")
-include("dsp_sipm_optimization.jl")
```
Comment (fhagemann) [inline src/LegendDSP.jl:43]: Why this?

Comment (apmypb) [reply src/LegendDSP.jl:43]: could be an artifact from main?

**Comment by apmypb** [comment]
What exactly do you mean by examples? Each filters has code examples in their docs
Or do you want unit tests?

**Comment by oschulz** [comment]
A test for each new filter would be great!

File: src/LegendDSP.jl (hunk 1 above)
Comment (fhagemann) [reply src/LegendDSP.jl:43]: This was introduced on `dev` 3 weeks ago in commit https://github.com/legend-exp/LegendDSP.jl/commit/4d2c0dc26f04c4aaf09cbb8c2b712971ef05c85b and should be kept

File: src/moving_window_multi.jl [hunk 2]
```diff
@@ -0,0 +1,118 @@
+import RadiationDetectorDSP: AbstractConvFilterInstance,
+    fltinstance, _smpllen, smplinfo, SamplingInfo, 
+    AbstractRadSigFilterInstance, LinearFiltering, _filterlen, _floattype
+
+"""
+    struct MovingWindowMulti <: AbstractRadFIRFilter
```
Comment (fhagemann) [inline src/moving_window_multi.jl:6]: Typo: missing `Filter` after `Multi`

File: src/ms_filter.jl [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+@inline _w(x, α) = exp(-α*x^2) + exp(-α*(x + 2)^2) + exp(-α*(x - 2)^2) - 2*exp(-α) - exp(-9*α)
+
+@inline κ_j(m::Int, v_n::AbstractArray) = v_n[1] + v_n[2]/(v_n[3] - m)^3
+
+"""
+    weighted_linear_reg(w::AbstractVector, y::AbstractVector)
+
+Do a weighted linear regression of the data `y` with weights `w` and return
+the offset and slope. its assumed that the
```
Comment (fhagemann) [inline src/ms_filter.jl:177]: its assumed that the... ?

File: src/moving_window_multi.jl [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+apply left and right moving average windows to signal.
+
```
Comment (fhagemann) [inline src/moving_window_multi.jl:9]: Could you also add a working example here?

File: src/ms_filter.jl [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+$(TYPEDFIELDS)
+"""
+Base.@kwdef struct ModifiedSincFilter{T<:RealQuantity} <: AbstractRadFIRFilter
+    "degree of the filter determining the number of extrema in the kernel"
+    d::Int = 2
+    "half-width of the kernel"
+    m::T = 3
+    function ModifiedSincFilter(d::Int, m::T) where T
+        s1 = "degree of ModifiedSincFilter needs to be even and be in the interval [2, 10]"
```
Comment (fhagemann) [inline src/ms_filter.jl:38]: use `D_MAX` in this string instead of `10`, in case we change `D_MAX`

File: src/ms_filter.jl [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    m::Int 
+    n_input::Int
+end
+
+Adapt.adapt_structure(to, flt::ModifiedSincFilter) = flt
+
+function fltinstance(flt::ModifiedSincFilter, fi::SamplingInfo{T}) where {T}
+    _m = round(Int, ustrip(NoUnits, flt.m/step(fi.axis)))
+    m_min = flt.d/2 + 2
+    s2 = "size of kernel to small for given degree"
```
Comment (fhagemann) [inline src/ms_filter.jl:62]: Typo: too

File: src/ms_filter.jl [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        @simd for j in eachindex(κ)
+            s = x*sin((2*j + ν)*π*x)
+            k_i = fma(κ[j], s, k_i)
+        end
+        k_i *= _w(x, α)
+        kernel[i] = k_i
+        sum += k_i
+        i > 1 && (sum += k_i)   # offcenter kernel elements appear twice
+    end
+    kernel ./= sum
```
Comment (fhagemann) [inline src/ms_filter.jl:128]: Did you check if this is correct?
I would naively say using `kernel ./= sum(kernel)` might be a good choice and get rid of the necessity to have the running `sum`.

File: src/ms_filter.jl [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    return kernel
+end
+
+"""
+    makeFitWeights(d::Int, m::Int)
+
+return the weights for the linear fit user for linear extrapolation at 
+at the right boundary.
+"""
+function makeFitWeights(d::U, m::Int) where U
```
Comment (fhagemann) [inline src/ms_filter.jl:138]: Looks like `d` can be anything according to the function but should be an `Int` according to the docstring. Does it make sense to constrain `U` to be `Real`, `Number` or `Integer`?

File: src/sgw_filter.jl [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+const SGW_COEFFS = Dict(
+    0 => [1., 1., -1.],                  # normal Sawitzky-Golay filter
+    1 => [0.68096, 0.36358, -3.68528],   # GAUSS2
+    2 => [0.67574, 0.35440, -3.61580],   # HANN
+    3 => [0.63944, 0.28417, -5.508],     # HANNSQR
+    4 => [0.62303, 0.25310, -7.07317]    # HANNCUBE
+)
+
+const WEIGHTTYPES = (0, 1, 2, 3, 4)
```
Comment (fhagemann) [inline src/sgw_filter.jl:68]: `const WEIGHTTYPES = keys(SGW_COEFFS)` in case more `SGW_COEFFS` are added?

File: src/sgw_filter.jl [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+function fltinstance(flt::WeightedSavitzkyGolayFilter, fi::SamplingInfo{T}
+    ) where {T}
+
+    fltlen = round(Int, ustrip(NoUnits, flt.length / step(fi.axis)))
```
Comment (fhagemann) [inline src/sgw_filter.jl:73]: The `ustrip(NoUnits, ...)` is used to really demand the `flt.length/step(fi.axis)` to have no units?

File: src/sgw_filter.jl [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+    m = (fltlen - 1) ÷ 2
+
+    l = length(fi.axis)
+    l_min = T(2*m + 1)
+    msg1 = "data too short; min length: $(l_min)"
+    @assert length(fi.axis) >= 2*m + 1 msg1
+    
+    d_max = 2*m
+    msg2 = "degree to big for given kernel size; max degree: $(d_max)"
```
Comment (fhagemann) [inline src/sgw_filter.jl:82]: Typo: too

File: src/sgw_filter.jl [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        pRight = ifelse((pRight + pLeft) > 2*m, 2*m - pLeft, pRight)
+        len_k = pRight + pLeft + 1
+        _unsafe_make_left_kernel!(kernel, P, weights, pLeft, len_k, 
+            scale, m, d, weightType)
+        for j in Base.OneTo(len_k)
+            idx = max(i-m-1, 0) + j
+            sum = fma(kernel[j], x[idx], sum)
+        end
+        y[i] = sum
+        sum = 0.
```
Comment (fhagemann) [inline src/sgw_filter.jl:130]: --> `sum = zero(T)`

File: src/sgw_filter.jl [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        len_k = pRight + pLeft + 1
+        _unsafe_make_left_kernel!(kernel, P, weights, pLeft, len_k, 
+            scale, m, d, weightType)
+        len_k = pRight + pLeft + 1
+        for j in Base.OneTo(len_k)
+            idx = L + 1 - j
+            sum = fma(kernel[j], x[idx], sum)
+        end
+        y[L-m+i] = sum
+        sum = 0.
```
Comment (fhagemann) [inline src/sgw_filter.jl:148]: --> `sum = zero(T)`

File: src/sgw_filter.jl [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+
+function _unsafe_dot_w(x::AbstractVector, y::AbstractVector, w::AbstractVector)
+    dotprod = 0.
+    @simd for i in eachindex(x)
+        dotprod = fma(x[i], fma(y[i], w[i], 0.), dotprod)
+    end
+    dotprod
+end
+
+function weightFunctionScale(missingFrac, weightType::Int)
```
Comment (fhagemann) [inline src/sgw_filter.jl:208]: Give `missingFrac` a type: `Real`?

File: src/sgw_filter.jl [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    coeffs = SGW_COEFFS[weightType]
+    missingFrac <= 0 ? 1 : _w(missingFrac, coeffs)
+end
+
+function weight_function(weight_type::Int, x::T) where T
+    decay = 2.0 # for GAUSS only
```
Comment (fhagemann) [inline src/sgw_filter.jl:214]: Then put this in the `weight_type == 1` block ;)

File: src/sgw_filter.jl [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+    if x <= -0.999999999999 || x >= 0.999999999999
+        return 0.0
```
Comment (fhagemann) [inline src/sgw_filter.jl:216]: `return zero(T)`?

File: src/sgw_filter.jl [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+    elseif weight_type == 0
+        return 1.0
```
Comment (fhagemann) [inline src/sgw_filter.jl:218]: `return one(T)`?

File: src/wh_filter.jl [hunk 18]
```diff
@@ -0,0 +1,147 @@
+"""
+
+    WhittakerHendersonFilter <: AbstractRadSigFilter{LinearFiltering}
```
Comment (fhagemann) [inline src/wh_filter.jl:3]: Everywhere else you write `struct` before, please keep it consistent

File: src/wh_filter.jl [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        sum = zero(T)
+        @simd for j in max(1, i-dmax):(i-1)
+            sum = fma(A[i-j+1, j], x[j], sum)
+        end
+        x[i] = (y[i] - sum) / A[1, i] 
+    end
+
+    # Backward substitution
+    @inbounds for i in reverse(axes(A, 2))
+        sum = 0.0
```
Comment (fhagemann) [inline src/wh_filter.jl:140]: --> `sum = zero(T)`

File: src/wh_filter.jl [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+            end
+        end
+    end
+    b
+end
+
+"""
+    solve!(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
+
+solve the lienar equation AA'x = y and store the result in x, where
```
Comment (fhagemann) [inline src/wh_filter.jl:120]: Typo: linear

File: src/wh_filter.jl [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+`A` is the cholesky decomposition of a banded centro symmetric matrix.
+A[1, i] is the diagonal, A[2, i] is the first subdiagonal and so on.
```
Comment (fhagemann) [inline src/wh_filter.jl:122]: In general: add ` around variables.

File: src/wh_filter.jl [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+"""
+function solve!(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
```
Comment (fhagemann) [inline src/wh_filter.jl:124]: Not sure if `solve!` is a good function for such a precise function. Might conflict with other packages.

File: src/ms_filter.jl [hunk 23]
```diff
@@ -0,0 +1,209 @@
+
+"""
+    struct ModifiedSincFilter{T<:RealQuantity} <: AbstractRadFIRFilter{}
+
+A [Modified-Sinc filter](https://doi.org/10.1021/acsmeasuresciau.3c00017)
```
Comment (fhagemann) [inline src/ms_filter.jl:5]: The links in the docstring direct to the Erratum, not the original article itself. Is that intended?

File: src/sgw_filter.jl (hunk 10 above)
Comment (apmypb) [reply src/sgw_filter.jl:73]: It also converts the units of `flt.length` and `step(fi.axis)` to a common unit before computing the fraction

**Review by fhagemann — APPROVED** [review]

File: src/LegendDSP.jl [hunk 24]
```diff
@@ -47,5 +47,9 @@ include("haar_filter.jl")
 include("ml.jl")
 include("dsp_ml_routines.jl")
 include("multi_intersect.jl")
+include("moving_window_multi.jl")
+include("ms_filter.jl")
```
Comment (theHenks) [inline src/LegendDSP.jl:51]: Can we put the filters `ms`, `sgw` and `wh` filter into one file? That would simplify the package strucutre to me.

File: test/runtests.jl [hunk 25]
```diff
@@ -8,5 +8,8 @@ Test.@testset "Package LegendDSP" begin
     # include("test_SOMETHING.jl")
     include("test_multiintersect.jl")
     include("test_docs.jl")
+    include("test_ms_filter.jl")
```
Comment (theHenks) [inline test/runtests.jl:11]: Same for the tests. Let's put them in one file for all three filters.

File: test/test_ms_filter.jl [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+    noise = 1.
+    t = range(0u"μs", 20u"μs", 2*n)
+    signal = vcat(zeros(n), 10*ones(n)) + (noise*rand(2*n) .- noise/2)
+    wf = RDWaveform(t, signal)
+    
+    # define filter parameters and filter
+    flt = ModifiedSincFilter(d=4, m=1u"μs")
+    
+    # apply filter to signal
+    @test_nowarn wf_new = flt(wf)
```
Comment (theHenks) [inline test/test_ms_filter.jl:18]: Maybe also a `@test isapprox(` style test at the end?

File: test/test_sgw_filter.jl [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+    noise = 1.
+    t = range(0u"μs", 20u"μs", 2*n)
+    signal = vcat(zeros(n), 10*ones(n)) + (noise*rand(2*n) .- noise/2)
+    wf = RDWaveform(t, signal)
+
+    # define filter parameters and filter
+    flt = WeightedSavitzkyGolayFilter(length=1u"μs", degree=3, weightType=2)
+
+    # apply filter to signal
+    @test_nowarn wf_new = flt(wf)
```
Comment (theHenks) [inline test/test_sgw_filter.jl:18]: Maybe also a `@test isapprox(` style test at the end?

File: test/test_wh_filter.jl [hunk 28]
```diff
@@ -0,0 +1,19 @@
...
+    noise = 1.
+    t = range(0u"μs", 20u"μs", 2*n)
+    signal = vcat(zeros(n), 10*ones(n)) + (noise*rand(2*n) .- noise/2)
+    wf = RDWaveform(t, signal)
+
+    # define filter parameters and filter
+    flt = WhittakerHendersonFilter(p=3, λ=4)
+
+    # apply filter to signal
+    @test_nowarn wf_new = flt(wf)
```
Comment (theHenks) [inline test/test_wh_filter.jl:18]: Maybe also a `@test isapprox(` style test at the end?

**Review by theHenks — CHANGES_REQUESTED** [review]
Fine to me. My two comments are:
- Let's move the filters into one file to make the package structure cleaner.
- Maybe the tests could also test with a `isapprox` style statement if the filter works.

**Review by fhagemann — APPROVED** [review]

**Comment by oschulz** [comment]
@apmypb could you make separate source/test files for the separate filters (without code duplication)?

**Comment by fhagemann** [comment]
> Fine to me. My two comments are:
> 
> * Let's move the filters into one file to make the package structure cleaner.

@oschulz this is asking for exactly the opposite as @theHenks ….

**Pull request merged**